amundsen_application/api/utils/metadata_utils.py

```diff
@@ -63,9 +65,50 @@ def marshall_table_full(table_dict: Dict) -> Dict:
     results['key'] = f'{table.database}://{table.cluster}.{table.schema}/{ table.name}'
     # Temp code to make 'partition_key' and 'partition_value' part of the table
     results['partition'] = _get_partition_data(results['watermarks'])
+
+    # We follow same style as column stat order for arranging the programmatic descriptions
+    prog_descriptions = results['programmatic_descriptions']
+    if prog_descriptions:
+        _update_prog_descriptions(prog_descriptions)
+
     return results
 
 
+def _update_prog_descriptions(prog_descriptions: List) -> None:
```

Comment:
I'd like to avoid habits of further processing  information for API calls, because any time the frontend service has to do this the immediate question is "For what reason does the API designed for this feature not return the exact information that frontend service needs for this feature". 

I recognize we have a need for a human-friendly display name in the UI for each of these section. However it is not a scalable choice for this to be something that has to be configured. (I know this pattern currently exists with badges but that was a workaround for the feature that will be difficult to maintain in environments with  many badges). Rather I think either:
1. Treat the title as piece of metadata itself, and it would be included in the payload returned from the metadata service. 
2. The React component that this displays this information would be responsible for transforming the `source` to a `title` according to rules for a particular pattern. If it becomes difficult to write logic for this, then to me would be that there needs to be some consistent patterns in how `source` is defined in the backend. As an example, see how we created a `Flag` component that has logic to determine how it should render it's text: https://github.com/lyft/amundsenfrontendlibrary/blob/master/amundsen_application/static/js/components/common/Flag/index.tsx#L18

I recognize that we intend to order this data. I'd like to understand more about why the list returned from metadata service needs to the data being ingested in such a ways that we can us the feature as is. I think about `columns` and the fact that those are passed to the frontend with a `sort_order`, and so I'm concerned abut diverging from that and having to maintain extra logic here.

Reply:
Ok thanks for that detailed writeup!

I decided not to include sort_order of programmatic_descriptions in the backend because the difference between sort_order of columns and sort_order of programmatic descriptions in my opinion is that one is actually real metadata and the other only pertains to presentation layer and has no actual real value. My thought was we should try and keep anything that isn't actually real metadata out of the backend if possible. This is also why "display_title" did not get put in the backend. Another example of something that followed this same design decision is column_stat_order, which is also configurable in config.py. currently.

Based on what you said, does this mean that column_stat_order would also be deprecated? I would imagine that column_stat_order would have a higher performance impact than programmatic_descriptions as it has to be performed for every column.

If these are all being deprecated, does this mean that we will be pushing all of this sort of "presentation configuration" in the backing db? Or is there another place that is recommended?

Here are a few options that I can think of, but I am sure that you guys have other ideas too:

1. Just remove it and only provide the default behavior for now (alphabetical sort, display title is just transformed (although is this more efficient than just a lookup in a config map?). Revisit in a different issue.
2. Backend changes I believe that would require that we normalize description sources and have a new "Description Source" node that would then contain these properties that could be easily changed without having to make changes to all the description nodes.
3. Relocate this configuration to the metadata_service? Not sure if thats any better.
4. Come up with a more efficient way to configure it on frontend?

In the meantime, I will look into at least modifying the display title like you mentioned for the Flag component.

Reply:
Column stat order would not be deprecated, but consider that programmatic descriptions are more like columns because they have no limit to their size. Meanwhile column stats is a fixed set  of values. So if we're grouping feature patterns based on characteristics, programmatic descriptions would be more akin to columns than the fixed set of column stats.

Another concept is that while I think there's value in separating what is metadata vs presentational information, we have to make design choices that are scalable. If we were dealing with a fixed set of values my thoughts would be different.

So now to focus on solutions:

1. I think I would be in favor of the alphabetical sort. Given that I liken these descriptions to columns, we do have a plan in our designs to put a sorting UI above the columns, which we could then put in for programmatic descriptions as well! However do you have use case for needing something other than alphabetical right now?

2. In terms of the display `title` being something that is transformed from the `source`, I'm currently not concerned about performance (given that it would take a LOT of data for this to have any noticeable affect on modern browsers). What I want to avoid is designing a feature where users will have to update the configuration every time they want to add a new description. Rather, if the information is ingested with certain rules in play -- such as the `source` following some common pattern like "snake_case" or "kebab-case"  -- then the frontend service will be able to transform new information for free.

Reply:
> What I want to avoid is designing a feature where users will have to update the configuration every time they want to add a new description. Rather, if the information is ingested with certain rules in play 
Oh! To clarify, the configuration is totally optional. By default, the programmatic_descriptions use the "description_source" as the title (I can make sure that it has the transform logic in place) and alphabetical sort. They don't need to be specified in the config file. These are just optional overrides. I put more information about that per @jornh suggestion in this code review in the docs/flask_config.py file. Maybe that will help clarify? If that is the case, would you be ok keeping the title overrides, or simpler to just get rid of all of it?

> Meanwhile column stats is a fixed set of values. 
Oh I thought that column stats were also unbounded as well?

> I think I would be in favor of the alphabetical sort. Given that I liken these descriptions to columns, we do have a plan in our designs to put a sorting UI above the columns, which we could then put in for programmatic descriptions as well! However do you have use case for needing something other than alphabetical right now?

Ok, we can definitely just keep it as only alphabetical sort for now.
In the future though, it would be nice to have control over the order of the descriptions because there are some reports we would like to be more immediately viewable, for example a quality badge about the dataset. Whereas, we have another job that populates some embedded graphs that can get quite long, so we would put that towards the end. I can't necessarily see it being useful to customize the sort differently in the UI for programmatic descriptions, but could be totally wrong about that!

Here is a screenshot that sort of demonstrates it, but not quite:

<img width="622" alt="Screen Shot 2020-03-11 at 3 45 17 PM" src="https://user-images.githubusercontent.com/2099463/76471251-67caec00-63af-11ea-868d-e6622c29771d.png">

Ideally we would want the charts to be at the end so that people don't have to scroll forever to get to other things. But it is not the end of the world and I am sure there is a better solution out there.

Reply:
I see. The configuration being optional makes things less painful and understanding that your use case deals with things like charts helps. I no longer think I mind having an optional sort order configuration, because I don't see a better way to address that use case. 

However for the display `title`, I currently still find it reasonable to expect some consistency in how what `source` is expected to look like. I think these sorts of consideration creates a more unified product, because in the worst case scenario if there is no consistency, then the optional configuration no longer becomes optional. 

We don't have to address everything in one go though, how about ping me other things are good to go and then we can re-evaluate. As a note if we do end up keeping the displayName configuration, with the existing precedents we would move that to the application configuration versus the Flask configuration. So if anything the Flask configuration would just have some concept of sort order and the application configuration would have the concept of the displayName